### DISCUSSION TITLE
How to use early_stopping_patience with deepAr? Also, some questions about the hyperparameters in the documentation

### DISCUSSION THREAD

**Discussion by Msaleh87** [awslabs/gluonts#2255] [Q&A] [answered]
So I tried to put early_stopping_patience = 6 in my code and it gave me a keyword error. 

```
estimator = DeepAREstimator(freq = '5min',
                           context_length = 2*24*12,  #The previous data points I look at to predict. It's like n_forecast in Neural Prophet 
                           prediction_length = 1*24*12,
                           num_layers = 2,
                           num_cells = 64,
                           cell_type = 'lstm', 
                           #cardinality = 'auto',
                           scaling = True,
                           use_feat_dynamic_real =True, 
                           early_stopping_patience = 6,  # <-----------
                           trainer = Trainer(epochs= 9, num_batches_per_epoch = 100, hybridize = True )) # increase the number of epochs
```
I have few questions regarding DeepAr parameters, I read this from Amazon: https://docs.aws.amazon.com/sagemaker/latest/dg/deepar_hyperparameters.html
"early_stopping_patience" was mentioned, but didn't say how
What exactly is hybridize = True?
What other cell_type I could use?
what is the difference between batch_size and num_batches_per_epoch?
Can the likelihood impact the actual forecast or only the confidence intervals? How to know which likelihood function to use, or is it based on trial, or are there cues I could look at?

**Comment by mbohlkeschneider** [comment, marked as answer]
Hi @Msaleh87,

Thank you for raising this issue. 

> I have few questions regarding DeepAr parameters, I read this from Amazon: https://docs.aws.amazon.com/sagemaker/latest/dg/deepar_hyperparameters.html
"early_stopping_patience" was mentioned, but didn't say how

This documentation refers to DeepAR in SageMaker as a first-party algorithm, which is a different implementation then the GluonTS DeepAR. Thus, this documentation does not apply to GluonTS. 

Here is a snipped to set the `patience` correctly for early stopping. It is a bit convoluted: 

```
from gluonts.model.deepar import DeepAREstimator
from gluonts.mx.trainer import Trainer
from gluonts.mx.trainer.learning_rate_scheduler import LearningRateReduction

learning_rate = 0.001

scheduler = LearningRateReduction(patience=20, base_lr=learning_rate, objective='min')
trainer = Trainer(learning_rate=learning_rate, callbacks=[scheduler])
est = DeepAREstimator(freq='H', prediction_length=1, trainer=trainer)
```

> What exactly is hybridize = True?

`hybridize=True` compiles the computation graph and runs the code in symbolic mode. You only need to ever switch this off for debugging purposes. 

> What other cell_type I could use?

Here are the available cell types: https://github.com/awslabs/gluon-ts/blob/dev/src/gluonts/mx/model/deepar/_estimator.py#L93

> what is the difference between batch_size and num_batches_per_epoch?

`batch_size` is the size of the batch and `num_batches_per_epoch` are the number of batches that are drawn for one epoch. Note that GluonTS does not use the standard notion of an epoch that is cycling to all data once but instead has a defined number of batches drawn. 

> Can the likelihood impact the actual forecast or only the confidence intervals? How to know which likelihood function to use, or is it based on trial, or are there cues I could look at?

You might decide from your data. For example, if you have count data you might want to use negative binomial. Other than that, it is trial and error.

**Reply by Msaleh87** [reply]
Thanks, @mbohlkeschneider, really appreciate your help!
I am using this documentation: https://ts.gluon.ai/stable/tutorials/forecasting/extended_tutorial.html#Configuring-an-estimator
If you know of any other please let me know.
I have a couple of questions:
When you said, "GluonTS does not use the standard notion of an epoch that is cycling to all data once but instead has a defined number of batches drawn." So what is the best practice when selecting the numbers for those two? Because obviously, if I try to increase the batch_size, the training time will be longer.
As for the Likelihood, here is what I am working on: comparing different tools/ algorithms to predict the electrical load of an electrical Microgird (a Microgrid, is a miniaturized version of the electric grid, it has solar panels, batteries, and could have other assets as well, and is connected to the main electric grid. It trades electricity with the grid based on the energy prices and could operate independently in case of grid failure). Now for the data-driven Microgrid controller to operate/ optimize the MG operation, cost-wise, it is essential and crucial to have a good accurate forecast of what is going to happen, so that the controller takes the forecast of the 24/48 hours ahead of the electric loads and solar panels and the rest of the resources and decide what would be the best action accordingly. A simple example, is during high prices, relaying as less as possible on the main electric grid and using the solar panels and batteries (or whatever assets it has).
I have been using FbProphet, Neural Prophet, DeepAr and tuning their hyperparameters to compare their forecast.
A sample of the data I have been dealing with is shown in the 2 pics below. I have data from March 2021 --> May 2022 with 1 min resolution, I resample it to 5 min. As you can see the electric load profile in 2021 was a bit noisy, but in 2022, it has more shape. Do you think from looking at these curves I could decide on the Likelihood?

Another question: Do you think data processing would help the forecast, like a Fast Fourier Transform or rolling average with a specific window? And can any of these data pre-processing techniques be included by a specific command within the estimator?


![Figure 10](https://user-images.githubusercontent.com/42920113/187771455-5a2dcb3f-17e3-425c-9c0d-cc5229c50c2f.png)

![Figure 11](https://user-images.githubusercontent.com/42920113/187771611-17adc624-1a0f-4e26-bb87-37b63575ca8b.png)

Thanks again for your time.

**Reply by mbohlkeschneider** [reply]
Hi @Msaleh87,

smoothing the data is usually not required. 

When I'm at a new problem with DeepAR, I rarely touch any parameters other than `context_length`, `epochs`, and `patience`. That usually does the trick. In many cases, a Student-t distribution works just fine.

**Reply by Msaleh87** [reply]
Hi @mbohlkeschneider, thanks for your help! I'll give it a try. One last question, I have a feature that I could only use with the training data set and not with the prediction, how can I input it, I know that dynamic_feat_real would not work. Should I use it as feat_static_real? The feature is basically some processing of the training data like a rolling average.

**Reply by mbohlkeschneider** [reply]
This is currently not implemented in DeepAR. I have [this PR](https://github.com/awslabs/gluon-ts/pull/1757) that implements that but I dropped the ball on it -_-. Having that merged would be required. Another method to explore would be [MQ-CNN](https://github.com/awslabs/gluon-ts/blob/dev/src/gluonts/mx/model/seq2seq/_mq_dnn_estimator.py#L38) for which this is already implemented.

**Reply by Msaleh87** [reply]
Ugh, I was worried that you would say that! I think that feature would have been really helpful! So, for now, this is not implemented in DeepAr? I was thinking to add dummy values in the prediction of feat_dynamic_rea, but I was worried it would mess up/ throw off my prediction.

As for MQ-CNN, I presume it's the same as DeepAr (coding-wise, but the concept is different) and if I want to use the feature, I should input it as: **past_feat_dynamic_real** and in the estimator add  **use_past_feat_dynamic_real = True**, right?

Does MQ-CNN has the same default for 5 min data as DeepAr? https://github.com/awslabs/gluon-ts/issues/2242#issuecomment-1232592602

**Reply by Msaleh87** [reply]
@mbohlkeschneider, is there any workaround to make DeepAr take unknown features in the future? I really hoped that the PR u referred to would have been implemented :)

**Comment by mbohlkeschneider** [comment]
Dummy values do not work well in my experience, so I would advise against that. The MQ-CNN flags you are pointing out are correct.

**Reply by Msaleh87** [reply]
@mbohlkeschneider I am using the following code:

```
from gluonts.model.seq2seq import MQCNNEstimator
estimator = MQCNNEstimator(freq = '5min',
                           #context_length = 2*24*12,  
                           prediction_length = 2*24*12,
                           scaling = True,
                           use_past_feat_dynamic_real = True,
                           trainer = Trainer(epochs= 10))
```
I was getting a memory error and I figured if I comment out the context_length and let the model figure it out on it's own it would work, and it did. However, in the first epoch, I see the estimated time to be about 7 hours! Is there something wrong I am doing and I missed it! Is that normal for CNN models? Also, if you have any recommendations for documentation of MQ-CNN, please share it

**Reply by Msaleh87** [reply]
Should I make the MQ-CNN a new question/ topic in the discussion?

**Comment by Msaleh87** [comment]
@mbohlkeschneider  Back to the original question regarding early_stopping_patience. I thought that its used to stop predictor in case there is no progress in the avg_epoch_loss, meaning: 
if the early_stopping_patience = 3 and the avg_epoch_loss has been the same for 3 epochs then the "estimator.train(training_data = training_data)" will stop. But, I read that if this is the case, no progress in the avg_epoch_loss, then the early_stopping_patience reduces the learning rate, which I think is the case.

I was confusing it with the early_stopping_patience from AWS: If this parameter is set, training stops when no progress is made within the specified number of epochs. The model that has the lowest loss is returned as the final model. https://docs.aws.amazon.com/sagemaker/latest/dg/deepar_hyperparameters.html

As for the patience code you provided, it reduces the learning rate in case there is no progress, correct? And if so, what is the best practice to use it? Is there any documentation regarding it? Because I noticed that every time I run the model with the same parameters, sometimes I would get totally different results and that was confusing, I don't know how much the patience code is reducing the learning rate
